### PR TITLE
added contact form Api    issue #16

### PR DESCRIPTION
**Pull request by AyushSharma72**
added a contact form Api which enable you to get user messages on mail

api

you will recieve messages like this on your mail id   mustafizkaifee04@gmail.com

check out the frontend repo modifications for this api   [frontend PR](https://github.com/GlobeHoppin/GlobeHoppin/pull/11)

### CODE REVIEW

File: api/controller/user.js [hunk 1]
```diff
@@ -12,3 +13,41 @@ export const getUserProfile = async (req, res) => {
...
+
+  const mailOptions = {
+    from: Email,
+    to: "mustafizkaifee04@gmail.com",
+    subject: `Message From Globe Hoppin`,
+    text: `
+      Email: ${Email}      
+      Subject:${Subject}
+      Message: ${Message}`,
+  };
```
Comment (Mustafiz04) [inline api/controller/user.js:42]: Better create a class for notification, add method `sendEmail()` take arguments then call this method from here.
Later on we can extend that class for further use case.

File: api/routes/user.js [hunk 2]
```diff
@@ -1,9 +1,11 @@
...
+import { getUserProfile, ContactAdmin } from "../controller";
 
-const UserRouter = new Express.Router()
+const UserRouter = new Express.Router();
 
-UserRouter.get('/', validateToken, getUserProfile)
+UserRouter.get("/", validateToken, getUserProfile);
 
-export default UserRouter
+UserRouter.post("/contact", ContactAdmin);
```
Comment (Mustafiz04) [inline api/routes/user.js:9]: This is not user specific api.

Create a new router `NotificationRouter` and new controller, new post api `/email` and controller method `sendEmail`

Right now we only support sending plain text message later on we can send sms, email with attachment, and different template for different emails.

File: app.js [hunk 3]
```diff
@@ -20,4 +25,4 @@ Routes.init(app);
 startServer(app);
 
 // For testing
-module.export = app;
+export default app; // Use export default instead of module.export
```
Comment (Mustafiz04) [inline app.js:28]: please remove this comment.

File: api/controller/user.js [hunk 4]
```diff
@@ -12,3 +13,41 @@ export const getUserProfile = async (req, res) => {
     return res.status(500).json({ error: "Internal Server Error" });
   }
 };
+
+export const ContactAdmin = async (req, resp) => {
+  if (req.method !== "POST") {
+    return resp.status(405).send("Method Not Allowed");
+  }
```
Comment (Mustafiz04) [inline api/controller/user.js:20]: Is this required ??

**Review by Mustafiz04 — CHANGES_REQUESTED** [review]

File: api/controller/user.js (hunk 4 above)
Comment (AyushSharma72) [reply api/controller/user.js:20]: i will remove it

**Comment by AyushSharma72** [comment]
please review the latest commit [commit](https://github.com/GlobeHoppin/globe-hoppin-backend/pull/17/commits/d34f2af553e186c92b5fbcff134d3516b27baf0f)

**Comment by Mustafiz04** [comment]
Can you please squash all 3 commits into 1 commit ?

**Comment by AyushSharma72** [comment]
> Can you please squash all 3 commits into 1 commit ?

i don't know how to do that is that necessary ?? if it is i will learn how to

**Comment by AyushSharma72** [comment]
> > Can you please squash all 3 commits into 1 commit ?
> 
> ok let me do it the rest is fine ??

please check @Mustafiz04

File: api/controller/notification.js [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        user: process.env.Email,
+        pass: process.env.Key,
+      },
+    });
+  }
+
+  async sendEmail({ email, subject, message }) {
+    const mailOptions = {
+      from: email,
+      to: "mustafizkaifee04+globehoppin@gmail.com",
```
Comment (Mustafiz04) [inline api/controller/notification.js:17]: We can move this to config.

File: api/controller/notification.js [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+      subject: `Message From Globe Hoppin`,
```
Comment (Mustafiz04) [inline api/controller/notification.js:18]: Subject will also change for different email. Pass subject as argument too.

File: api/routes/notification.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import express from "express";
+import { validateToken } from "../middleware/auth";
+import NotificationService from "../controller/notification"; //  the NotificationService class
```
Comment (Mustafiz04) [inline api/routes/notification.js:3]: Can remove this comment.

File: api/routes/notification.js [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+const NotificationRouter = express.Router();
+const notificationService = new NotificationService();
+
+// Route handler
+NotificationRouter.post("/email", async (req, res) => {
+  const { email, subject, message } = req.body;
```
Comment (Mustafiz04) [inline api/routes/notification.js:10]: Always initialize with default value while destructuring.
```suggestion
  const { body = {} } = req; 
  const { email= '', subject = '', message = '' } = body;
```

Comment (Mustafiz04) [reply api/routes/notification.js:10]: Also please add validation on these field.

File: api/controller/notification.js (hunk 5 above)
Comment (AyushSharma72) [reply api/controller/notification.js:17]: > We can move this to config.

please explain

File: api/controller/notification.js (hunk 6 above)
Comment (AyushSharma72) [reply api/controller/notification.js:18]: > Subject will also change for different email. Pass subject as argument too.

subject is already a field this is the title of the email

image

**Comment by Mustafiz04** [comment]
Please squash and write the correct commit message.
I am asking this so it helps us track work correctly.

**Comment by AyushSharma72** [comment]
> Please squash and write the correct commit message. I am asking this so it helps us track work correctly.

yes was about to do thaty

File: api/controller/notification.js (hunk 6 above)
Comment (Mustafiz04) [reply api/controller/notification.js:18]: Yes, that subject is the actual subject of email, and the one in text is email body.
```suggestion
      subject,
```

**Review by Mustafiz04 — CHANGES_REQUESTED** [review]
Just one change, rest look good.

**Comment by AyushSharma72** [comment]
> Just one change, rest look good.

ok please wait so i will make the subject to the email title and  remove it form the fields ??then we will have only two fields the email and the messages ??

**Comment by Mustafiz04** [comment]
> ok please wait so i will make the subject to the email title and remove it form the fields ??then we will have only two fields the email and the messages ??

We should only have a message. Right now your code sends emails in plain text, you can also explore sending an email template using EJS or something. It will send emails professionally and look good. I hope you got what I am saying.
https://ejs.co/

**Comment by AyushSharma72** [comment]
> > ok please wait so i will make the subject to the email title and remove it form the fields ??then we will have only two fields the email and the messages ??
> 
> We should only have a message. Right now your code sends emails in plain text, you can also explore sending an email template using EJS or something. It will send emails professionally and look good. I hope you got what I am saying. https://ejs.co/

what should i do now ? remove the subject right 
ejs and template i will explore later sir may be in other issue thankyou for sharing your experience

**Comment by Mustafiz04** [comment]
Follow this to squash commits -- https://github.com/GlobeHoppin/globe-hoppin-backend?tab=readme-ov-file#note

**Comment by Mustafiz04** [comment]
> > > ok please wait so i will make the subject to the email title and remove it form the fields ??then we will have only two fields the email and the messages ??
> > 
> > 
> > We should only have a message. Right now your code sends emails in plain text, you can also explore sending an email template using EJS or something. It will send emails professionally and look good. I hope you got what I am saying. https://ejs.co/
> 
> what should i do now ? remove the subject right ejs and template i will explore later sir may be in other issue thankyou for sharing your experience

Yeah remove the subject and create a follow-up issue to use the EJS template and work on it.

**Comment by AyushSharma72** [comment]
> > > > ok please wait so i will make the subject to the email title and remove it form the fields ??then we will have only two fields the email and the messages ??
> > > 
> > > 
> > > We should only have a message. Right now your code sends emails in plain text, you can also explore sending an email template using EJS or something. It will send emails professionally and look good. I hope you got what I am saying. https://ejs.co/
> > 
> > 
> > what should i do now ? remove the subject right ejs and template i will explore later sir may be in other issue thankyou for sharing your experience
> 
> Yeah remove the subject and create a follow-up issue to use the EJS template and work on it.

before that will you merge this please ??

**Comment by AyushSharma72** [comment]
done ! please merge now 🙏🙏also check the forntend code

**Comment by Mustafiz04** [comment]
Everything look good but please merge both commit into one. There are still 2 commits showing.

Let me know if you are not able to squash.

It is good practice to have only 1 commit per PR.

**Comment by AyushSharma72** [comment]
> Everything look good but please merge both commit into one. There are still 2 commits showing.
> 
> Let me know if you are not able to squash.
> 
> It is good practice to have only 1 commit per PR.

I will try again

**Comment by AyushSharma72** [comment]
> > Everything look good but please merge both commit into one. There are still 2 commits showing.
> > Let me know if you are not able to squash.
> > It is good practice to have only 1 commit per PR.
> 
> I will try again

completed only one commit now

**Review by Mustafiz04 — APPROVED** [review]

**Comment by AyushSharma72** [comment]
How this will work with out frontend please check GlobeHoppin/GlobeHoppin#11

**Pull request merged**